session.go

```diff
@@ -398,6 +448,9 @@ func (s *session) onSessionError(readErr error, writeErr error) {
 		// considered no good at this point and we won't bother sending anything.
 		c.close(false, readErr, writeErr)
 	}
+	s.closeOnce.Do(func() {
+		close(s.out)
+	})
```

Comment:
It kind of seems like all of the `onSessionError` stuff should just be in the `Close` method, partly since AFAICT it's the only thing that calls close, but I don't think it's a huge deal...

Reply:
Do we also want to call `close(s.echoOut)` here and also correspondingly handle `s.echoOut` being closed @oxtoacart?

Reply:
As you noted, session is a net.Conn. Whoever is using it does in fact call `Close()`, but not because we're handling an error. I do think this can be unified a bit better, give me a few.

Reply:
Oh, scratch that. After messing with this, all I did was break the unit test. I'm leaving as-is.

Reply:
OK

Reply:
Concerning echoOut, we don't need to close that. The reason that we close `s.out` is to interrupt the `sendLoop`. We read from both `s.out` and `s.echoOut` in the same select in that loop, so `s.out` being closed is enough to stop us here.

See [this example](https://play.golang.org/p/vp_JFyT9MK).